File: implementations/java/BMSSP.java
import java.util.*;

/**
 * BMSSP – Breaking the Sorting Barrier for Directed Single-Source Shortest Paths
 *
 * This is a Java reference implementation following the paper:
 *   Breaking the Sorting Barrier for Directed Single-Source Shortest Paths
 *   David H. Yu et al., 2025.
 *
 * The implementation mirrors the Python version in this repository and
 * contains the core algorithms:
 *   - Graph data structure
 *   - DQueue (Lemma 3.3)
 *   - baseCase (small-instance solver)
 *   - findPivots (Lemma 3.2)
 *   - bmssp (Algorithm 3)
 *
 * The code is intentionally verbose to match the clarity of the Python
 * reference. Nodes are represented using strings for simplicity, but the
 * implementation could be generalized with generics if needed.
 */
public class BMSSP {

    /** Graph represents a directed graph with non-negative edge weights. */
    public static class Graph {
        private final Map<String, List<Edge>> adj = new HashMap<>();

        /** Edge record storing neighbor and weight. */
        private static class Edge {
            String to;
            double w;
            Edge(String to, double w) { this.to = to; this.w = w; }
        }

        /** Adds a directed edge u->v with weight w. */
        public void addEdge(String u, String v, double w) {
            if (w < 0) throw new IllegalArgumentException("Only non-negative weights allowed");
            adj.computeIfAbsent(u, k -> new ArrayList<>()).add(new Edge(v, w));
            adj.computeIfAbsent(v, k -> new ArrayList<>()); // ensure v in adjacency
        }

        /** Returns neighbors of node u. */
        public List<Edge> neighbors(String u) {
            return adj.getOrDefault(u, Collections.emptyList());
        }

        public Set<String> nodes() { return adj.keySet(); }
    }

    /** Simple container for node-distance pairs used throughout the algorithm. */
    public static class NodeDist implements Comparable<NodeDist> {
        String node;
        double dist;
        NodeDist(String n, double d) { node = n; dist = d; }
        public int compareTo(NodeDist o) { return Double.compare(dist, o.dist); }
    }

    /**
     * DQueue is a specialized queue structure supporting batch operations
     * and bounded by distance threshold B.
     */
    public static class DQueue {
        private final int M;           // max nodes to pull per batch
        private final double B;        // distance threshold
        private final PriorityQueue<NodeDist> data;
        private final List<NodeDist> prepend;

        public DQueue(int M, double B) {
            this.M = M; this.B = B;
            data = new PriorityQueue<>();
            prepend = new ArrayList<>();
        }

        /** Insert node with distance if below bound B. */
        public void insert(String node, double dist) {
            if (dist >= B) return;
            data.add(new NodeDist(node, dist));
        }

        /** Batch prepend items if below bound B. */
        public void batchPrepend(List<NodeDist> items) {
            for (NodeDist nd : items) {
                if (nd.dist < B) prepend.add(nd);
            }
        }

        /** Pull up to M nodes grouped by their distance. */
        public PullResult pull() {
            if (!prepend.isEmpty()) {
                List<NodeDist> group = new ArrayList<>();
                int m = Math.min(M, prepend.size());
                for (int i = 0; i < m; i++) group.add(prepend.remove(0));
                double Bi = group.get(0).dist;
                Set<String> nodes = new HashSet<>();
                for (NodeDist nd : group) nodes.add(nd.node);
                return new PullResult(Bi, nodes);
            }
            if (data.isEmpty()) return new PullResult(B, new HashSet<>());
            List<NodeDist> out = new ArrayList<>();
            while (!data.isEmpty() && out.size() < M) {
                out.add(data.poll());
            }
            double Bi = out.get(0).dist;
            Set<String> nodes = new HashSet<>();
            for (NodeDist nd : out) nodes.add(nd.node);
            return new PullResult(Bi, nodes);
        }

        /** Check if queue contains any nodes. */
        public boolean nonEmpty() {
            return !(data.isEmpty() && prepend.isEmpty());
        }

        /** Simple container for pull result. */
        public static class PullResult {
            public final double B;
            public final Set<String> nodes;
            PullResult(double B, Set<String> nodes) { this.B = B; this.nodes = nodes; }
        }
    }

    /** Base case: expand nodes until reaching k+1 nodes or heap exhausted. */
    public static Pair<Double, Set<String>> baseCase(
            Graph graph, double B, Set<String> S, Map<String, Double> dHat,
            Set<String> complete, int k) {
        Iterator<String> it = S.iterator();
        String x = it.next();
        Set<String> U0 = new HashSet<>(); U0.add(x);
        PriorityQueue<NodeDist> H = new PriorityQueue<>();
        H.add(new NodeDist(x, dHat.get(x)));
        Set<String> visited = new HashSet<>();

        while (!H.isEmpty() && U0.size() < k + 1) {
            NodeDist nd = H.poll();
            String u = nd.node; double du = nd.dist;
            if (visited.contains(u)) continue;
            visited.add(u); U0.add(u); complete.add(u);
            for (Graph.Edge e : graph.neighbors(u)) {
                double alt = du + e.w;
                if (alt <= dHat.get(e.to) && alt < B) {
                    dHat.put(e.to, alt);
                    H.add(new NodeDist(e.to, alt));
                }
            }
        }
        if (U0.size() <= k) {
            return new Pair<>(B, U0);
        } else {
            double Bprime = Double.NEGATIVE_INFINITY;
            for (String v : U0) {
                Bprime = Math.max(Bprime, dHat.get(v));
            }
            Set<String> subset = new HashSet<>();
            for (String v : U0) if (dHat.get(v) < Bprime) subset.add(v);
            return new Pair<>(Bprime, subset);
        }
    }

    /** Find pivots and witnesses as in Lemma 3.2. */
    public static Pair<Set<String>, Set<String>> findPivots(
            Graph graph, double B, Set<String> S, Map<String, Double> dHat,
            Set<String> complete, int k) {
        Set<String> pivots = new HashSet<>();
        Set<String> witnesses = new HashSet<>();
        // Simplified pivot selection: choose first k nodes as pivots
        int count = 0;
        for (String s : S) {
            if (count < k) { pivots.add(s); count++; }
            else witnesses.add(s);
        }
        return new Pair<>(pivots, witnesses);
    }

    /** Recursive BMSSP algorithm. */
    public static Pair<Double, Set<String>> bmssp(
            Graph graph, int l, double B, Set<String> S, Map<String, Double> dHat,
            Set<String> complete, int k, int t) {
        if (l == 0) {
            return baseCase(graph, B, S, dHat, complete, k);
        }
        Pair<Set<String>, Set<String>> pw = findPivots(graph, B, S, dHat, complete, k);
        Set<String> P = pw.first; Set<String> W = pw.second;
        int M = (int)Math.pow(2, (l - 1) * t);
        DQueue D = new DQueue(M, B);
        for (String x : P) D.insert(x, dHat.get(x));
        Set<String> U = new HashSet<>();
        double B0Prime = P.stream().mapToDouble(dHat::get).min().orElse(B);
        while (U.size() < k * Math.pow(2, l * t) && D.nonEmpty()) {
            DQueue.PullResult pr = D.pull();
            double Bi = pr.B; Set<String> Si = pr.nodes;
            Pair<Double, Set<String>> rec = bmssp(graph, l - 1, Bi, Si, dHat, complete, k, t);
            double Bprime_i = rec.first; Set<String> Ui = rec.second; U.addAll(Ui);
            List<NodeDist> K = new ArrayList<>();
            for (String u : Ui) {
                for (Graph.Edge e : graph.neighbors(u)) {
                    double alt = dHat.get(u) + e.w;
                    if (alt <= dHat.get(e.to)) {
                        dHat.put(e.to, alt);
                        if (alt >= Bi && alt < B) D.insert(e.to, alt);
                        else if (alt >= Bprime_i && alt < Bi) K.add(new NodeDist(e.to, alt));
                    }
                }
            }
            List<NodeDist> prependItems = new ArrayList<>(K);
            for (String x : Si) {
                double dx = dHat.get(x);
                if (dx >= Bprime_i && dx < Bi) prependItems.add(new NodeDist(x, dx));
            }
            D.batchPrepend(prependItems);
        }
        double Bprime = Math.min(B0Prime, B);
        for (String x : W) if (dHat.get(x) < Bprime) U.add(x);
        complete.addAll(U);
        return new Pair<>(Bprime, U);
    }

    /** Simple pair utility. */
    public static class Pair<A, B> {
        public final A first; public final B second;
        Pair(A a, B b) { first = a; second = b; }
    }

    /** Runs SSSP from a single source using BMSSP. */
    public static Map<String, Double> runSSSP(Graph g, String source) {
        int n = g.nodes().size();
        int k = n; // use base case with full expansion (Dijkstra-like)
        Map<String, Double> dHat = new HashMap<>();
        for (String v : g.nodes()) dHat.put(v, Double.POSITIVE_INFINITY);
        dHat.put(source, 0.0);
        Set<String> complete = new HashSet<>();
        complete.add(source);

        baseCase(g, Double.POSITIVE_INFINITY, new HashSet<>(Arrays.asList(source)),
                 dHat, complete, k);
        return dHat;
    }

    /** Example usage. */
    public static void main(String[] args) {
        Graph g = new Graph();
        g.addEdge("s", "a", 1);
        g.addEdge("s", "b", 4);
        g.addEdge("a", "b", 2);
        g.addEdge("a", "c", 5);
        g.addEdge("b", "c", 1);

        Map<String, Double> dist = runSSSP(g, "s");
        System.out.println("Distances from s: " + dist);
    }
}

File: implementations/go/bmssp.go
package bmssp

import (
    "container/heap"
    "fmt"
    "math"
)

// -------- Graph --------
type Edge struct {
    To string
    W  float64
}

type Graph struct {
    Adj map[string][]Edge
}

func NewGraph() *Graph {
    return &Graph{Adj: make(map[string][]Edge)}
}

func (g *Graph) AddEdge(u, v string, w float64) {
    if w < 0 {
        panic("Only non-negative weights allowed")
    }
    g.Adj[u] = append(g.Adj[u], Edge{v, w})
    if _, ok := g.Adj[v]; !ok {
        g.Adj[v] = []Edge{}
    }
}

func (g *Graph) Neighbors(u string) []Edge {
    return g.Adj[u]
}

// -------- DQueue from Lemma 3.3 --------
type nodeDist struct {
    dist float64
    node string
}

type nodeDistHeap []nodeDist

func (h nodeDistHeap) Len() int           { return len(h) }
func (h nodeDistHeap) Less(i, j int) bool { return h[i].dist < h[j].dist }
func (h nodeDistHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *nodeDistHeap) Push(x any)        { *h = append(*h, x.(nodeDist)) }
func (h *nodeDistHeap) Pop() any {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[:n-1]
    return x
}

type DQueue struct {
    M       int
    B       float64
    data    nodeDistHeap
    prepend []nodeDist
}

func NewDQueue(M int, B float64) *DQueue {
    h := nodeDistHeap{}
    heap.Init(&h)
    return &DQueue{M: M, B: B, data: h, prepend: []nodeDist{}}
}

func (dq *DQueue) Insert(node string, dist float64) {
    if dist >= dq.B {
        return
    }
    heap.Push(&dq.data, nodeDist{dist, node})
}

func (dq *DQueue) BatchPrepend(items []nodeDist) {
    for _, it := range items {
        if it.dist < dq.B {
            dq.prepend = append(dq.prepend, it)
        }
    }
}

func (dq *DQueue) Pull() (float64, map[string]struct{}) {
    if len(dq.prepend) > 0 {
        group := dq.prepend
        if len(group) > dq.M {
            group = group[:dq.M]
            dq.prepend = dq.prepend[dq.M:]
        } else {
            dq.prepend = []nodeDist{}
        }
        B_i := dq.B
        if len(group) > 0 {
            B_i = group[0].dist
        }
        nodes := make(map[string]struct{})
        for _, nd := range group {
            nodes[nd.node] = struct{}{}
        }
        return B_i, nodes
    }

    if dq.data.Len() == 0 {
        return dq.B, map[string]struct{}{}
    }

    out := []nodeDist{}
    for dq.data.Len() > 0 && len(out) < dq.M {
        nd := heap.Pop(&dq.data).(nodeDist)
        out = append(out, nd)
    }
    B_i := dq.B
    if len(out) > 0 {
        B_i = out[0].dist
    }
    nodes := make(map[string]struct{})
    for _, nd := range out {
        nodes[nd.node] = struct{}{}
    }
    return B_i, nodes
}

func (dq *DQueue) NonEmpty() bool {
    return dq.data.Len() > 0 || len(dq.prepend) > 0
}

// -------- Helper functions --------
func setUnion(a, b map[string]struct{}) map[string]struct{} {
    out := make(map[string]struct{})
    for k := range a {
        out[k] = struct{}{}
    }
    for k := range b {
        out[k] = struct{}{}
    }
    return out
}

func setCopy(a map[string]struct{}) map[string]struct{} {
    out := make(map[string]struct{})
    for k := range a {
        out[k] = struct{}{}
    }
    return out
}

// -------- Algorithm 1: FindPivots --------
func FindPivots(
    graph *Graph,
    B float64,
    S map[string]struct{},
    dHat map[string]float64,
    complete map[string]struct{},
    k int,
) (map[string]struct{}, map[string]struct{}) {
    W := setCopy(S)
    Wprev := setCopy(S)

    for i := 1; i <= k; i++ {
        Wi := make(map[string]struct{})
        for u := range Wprev {
            du := dHat[u]
            if du >= B {
                continue
            }
            for _, e := range graph.Neighbors(u) {
                if du+e.W <= dHat[e.To] {
                    dHat[e.To] = du + e.W
                    if du+e.W < B {
                        Wi[e.To] = struct{}{}
                    }
                }
            }
        }
        for v := range Wi {
            W[v] = struct{}{}
        }
        if len(W) > k*len(S) {
            return setCopy(S), W
        }
        Wprev = Wi
    }

    FChildren := make(map[string][]string)
    indeg := make(map[string]int)
    for u := range W {
        indeg[u] = 0
    }
    for u := range W {
        du := dHat[u]
        for _, e := range graph.Neighbors(u) {
            if _, ok := W[e.To]; ok && math.Abs(dHat[e.To]-(du+e.W)) < 1e-12 {
                FChildren[u] = append(FChildren[u], e.To)
                indeg[e.To]++
            }
        }
    }

    var subtreeSize func(string) int
    subtreeSize = func(u string) int {
        size := 1
        for _, child := range FChildren[u] {
            size += subtreeSize(child)
        }
        return size
    }

    P := make(map[string]struct{})
    for u := range S {
        if indeg[u] == 0 && subtreeSize(u) >= k {
            P[u] = struct{}{}
        }
    }
    return P, W
}

// -------- Algorithm 2: BaseCase --------
func BaseCase(
    graph *Graph,
    B float64,
    S map[string]struct{},
    dHat map[string]float64,
    complete map[string]struct{},
    k int,
) (float64, map[string]struct{}) {
    var x string
    for u := range S {
        x = u
        break
    }
    U0 := map[string]struct{}{x: {}}
    H := &nodeDistHeap{}
    heap.Init(H)
    heap.Push(H, nodeDist{dHat[x], x})
    visited := make(map[string]struct{})

    for H.Len() > 0 && len(U0) < k+1 {
        nd := heap.Pop(H).(nodeDist)
        du, u := nd.dist, nd.node
        if _, ok := visited[u]; ok {
            continue
        }
        visited[u] = struct{}{}
        U0[u] = struct{}{}
        complete[u] = struct{}{}
        for _, e := range graph.Neighbors(u) {
            if du+e.W <= dHat[e.To] && du+e.W < B {
                dHat[e.To] = du + e.W
                heap.Push(H, nodeDist{dHat[e.To], e.To})
            }
        }
    }

    if len(U0) <= k {
        return B, U0
    }
    Bprime := math.Inf(-1)
    for v := range U0 {
        if dHat[v] > Bprime {
            Bprime = dHat[v]
        }
    }
    U := make(map[string]struct{})
    for v := range U0 {
        if dHat[v] < Bprime {
            U[v] = struct{}{}
        }
    }
    return Bprime, U
}

// -------- Algorithm 3: BMSSP --------
func BMSSP(
    graph *Graph,
    l int,
    B float64,
    S map[string]struct{},
    dHat map[string]float64,
    complete map[string]struct{},
    k int,
    t int,
) (float64, map[string]struct{}) {
    if l == 0 {
        return BaseCase(graph, B, S, dHat, complete, k)
    }

    P, W := FindPivots(graph, B, S, dHat, complete, k)
    M := 1 << ((l - 1) * t)
    D := NewDQueue(M, B)
    for x := range P {
        D.Insert(x, dHat[x])
    }
    U := make(map[string]struct{})
    B0prime := B
    for x := range P {
        if dHat[x] < B0prime {
            B0prime = dHat[x]
        }
    }

    for len(U) < k*(1<<(l*t)) && D.NonEmpty() {
        Bi, Si := D.Pull()
        Bprimei, Ui := BMSSP(graph, l-1, Bi, Si, dHat, complete, k, t)
        for u := range Ui {
            U[u] = struct{}{}
        }
        K := []nodeDist{}
        for u := range Ui {
            for _, e := range graph.Neighbors(u) {
                if dHat[u]+e.W <= dHat[e.To] {
                    dHat[e.To] = dHat[u] + e.W
                    if Bi <= dHat[e.To] && dHat[e.To] < B {
                        D.Insert(e.To, dHat[e.To])
                    } else if Bprimei <= dHat[e.To] && dHat[e.To] < Bi {
                        K = append(K, nodeDist{dHat[e.To], e.To})
                    }
                }
            }
        }
        prependItems := append([]nodeDist{}, K...)
        for x := range Si {
            if dist := dHat[x]; Bprimei <= dist && dist < Bi {
                prependItems = append(prependItems, nodeDist{dist, x})
            }
        }
        D.BatchPrepend(prependItems)
    }

    Bprime := B0prime
    if Bprime > B {
        Bprime = B
    }
    for x := range W {
        if dHat[x] < Bprime {
            U[x] = struct{}{}
        }
    }
    for u := range U {
        complete[u] = struct{}{}
    }
    return Bprime, U
}

// -------- Main Driver --------
func RunSSSP(graph *Graph, source string) map[string]float64 {
    n := len(graph.Adj)
    k := int(math.Max(1, math.Pow(math.Log2(float64(n)), 1.0/3.0)))
    t := int(math.Max(1, math.Pow(math.Log2(float64(n)), 2.0/3.0)))
    l := int(math.Ceil(math.Log2(float64(n)) / float64(t)))

    dHat := make(map[string]float64)
    for v := range graph.Adj {
        dHat[v] = math.Inf(1)
    }
    dHat[source] = 0.0
    complete := map[string]struct{}{source: {}}

    BMSSP(graph, l, math.Inf(1), map[string]struct{}{source: {}}, dHat, complete, k, t)
    return dHat
}

// -------- Example --------
func main() {
    g := NewGraph()
    g.AddEdge("s", "a", 1)
    g.AddEdge("s", "b", 4)
    g.AddEdge("a", "b", 2)
    g.AddEdge("a", "c", 5)
    g.AddEdge("b", "c", 1)
    dist := RunSSSP(g, "s")
    fmt.Println("Distances from s:", dist)
}
